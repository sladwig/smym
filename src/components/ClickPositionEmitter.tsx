import React from 'react';
import './ClickPositionEmitter.css';

interface IProps {
    value?: string;
    onClick?: (position: number) => any;
    startPosition?: number;
}

export const ClickPositionEmitter = ({
    value,
    onClick = (_: number) => null,
    startPosition = 0,
}: IProps) => {
    if (value === undefined) return null;
    const chars = value.split('').map((char, index) => {
        const position = startPosition + index;
        return (
            <PositionEmitter
                key={`char-${position}`}
                value={char}
                position={position}
                onClick={position => onClick(position)}
            />
        );
    });

    return <span className="click-position-emitter">{chars}</span>;
};

interface IPositionEmitterProps {
    value?: string;
    position?: number;
    onClick?: (position: number) => any;
}
export const PositionEmitter = ({
    value,
    position = 0,
    onClick = (_: number) => null,
}: IPositionEmitterProps) => {
    const sendPosition = (pos: number) => (event: React.MouseEvent) => {
        if (event && event.preventDefault) event.preventDefault();
        if (event && event.stopPropagation) event.stopPropagation();
        onClick(pos);
    };
    return (
        <span className="position-emitter">
            {value}
            <span className="position-area-overlay">
                <span className="position-area" onClick={sendPosition(position)}></span>
                <span className="position-area" onClick={sendPosition(position + 1)}></span>
            </span>
        </span>
    );
};
