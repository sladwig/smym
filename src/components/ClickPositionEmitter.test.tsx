import { render, fireEvent } from '@testing-library/react';
import React from 'react';
import { ClickPositionEmitter, PositionEmitter } from './ClickPositionEmitter';

describe('ClickPositionEmitter', () => {
    test('renders', () => {
        const { container } = render(<ClickPositionEmitter />);
        expect(container).toMatchSnapshot();
    });

    test('renders with value', () => {
        const { container } = render(<ClickPositionEmitter value="abc" />);
        expect(container).toMatchSnapshot();
    });

    test('is clickable and emits click position', () => {
        const mock = jest.fn();
        const { container } = render(<ClickPositionEmitter value={'ab'} onClick={mock} />);
        container.querySelectorAll('.position-area').forEach(area => {
            fireEvent.click(area);
        });
        expect(mock.mock.calls).toEqual([[0], [1], [1], [2]]);
    });

    test('can have a start position', () => {
        const mock = jest.fn();
        const { container } = render(
            <ClickPositionEmitter value={'ab'} onClick={mock} startPosition={3} />,
        );
        container.querySelectorAll('.position-area').forEach(area => {
            fireEvent.click(area);
        });
        expect(mock.mock.calls).toEqual([[3], [4], [4], [5]]);
    });
});

describe('PositionEmitter', () => {
    test('renders', () => {
        const { container } = render(<PositionEmitter value="a" />);
        expect(container).toMatchSnapshot();
    });

    test('is clickable and emits click position', () => {
        const mock = jest.fn();
        const { container } = render(<PositionEmitter value={'ab'} onClick={mock} />);
        container.querySelectorAll('.position-area').forEach(area => {
            fireEvent.click(area);
        });
        expect(mock.mock.calls).toEqual([[0], [1]]);
    });

    test('can set position', () => {
        const mock = jest.fn();
        const { container } = render(<PositionEmitter value={'ab'} onClick={mock} position={3} />);
        container.querySelectorAll('.position-area').forEach(area => {
            fireEvent.click(area);
        });
        expect(mock.mock.calls).toEqual([[3], [4]]);
    });
});
