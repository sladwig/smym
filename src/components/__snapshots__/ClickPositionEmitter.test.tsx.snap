// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ClickPositionEmitter renders 1`] = `<div />`;

exports[`ClickPositionEmitter renders with value 1`] = `
<div>
  <span
    class="click-position-emitter"
  >
    <span
      class="position-emitter"
    >
      a
      <span
        class="position-area-overlay"
      >
        <span
          class="position-area"
        />
        <span
          class="position-area"
        />
      </span>
    </span>
    <span
      class="position-emitter"
    >
      b
      <span
        class="position-area-overlay"
      >
        <span
          class="position-area"
        />
        <span
          class="position-area"
        />
      </span>
    </span>
    <span
      class="position-emitter"
    >
      c
      <span
        class="position-area-overlay"
      >
        <span
          class="position-area"
        />
        <span
          class="position-area"
        />
      </span>
    </span>
  </span>
</div>
`;

exports[`PositionEmitter renders 1`] = `
<div>
  <span
    class="position-emitter"
  >
    a
    <span
      class="position-area-overlay"
    >
      <span
        class="position-area"
      />
      <span
        class="position-area"
      />
    </span>
  </span>
</div>
`;
